File: src/screens/SettingsScreen.lua
-- Skelly settings screen.
--
-- By Chris Herborth (https://github.com/Taffer)
-- MIT license, see LICENSE.md for details.

local Class = require 'lib/middleclass/middleclass'
local ScreenBase = require 'src/screens/ScreenBase'

local ColorFade = require 'src/ColorFade'
local ImageButton = require 'src/ui/ImageButton'
local Label = require 'src/ui/Label'
local SettingsOverlay = require 'src/ui/SettingsOverlay'

local SettingsScreen = Class('SettingsScreen', ScreenBase)

function SettingsScreen:initialize(resources, state)
    ScreenBase.initialize(self, resources, state)
    self:setNextScreen('Journey')

    self.skelly_text = self.resources.text.skelly_title
    self.subtitle_text = self.resources.text.title.subtitle_text

    self.fade = ColorFade:new({0, 0, 0, 1}, {0, 0, 0, 0}, 1)

    local title_image = self.resources.images.skelly_title
    local title_quad = love.graphics.newQuad(0, 0, title_image:getWidth(), title_image:getHeight(), title_image)

    local font_mono = self.resources.fonts.default_mono
    local font_title = self.resources.fonts.skelly_title

    self.ui = {
        ImageButton:new(self, 0, 0, title_image, title_quad),
        Label:new(self, state.scr_width / 2, 40, self.skelly_text, font_title, {1, 1, 1, 1}, 'centre'),
        Label:new(self, state.scr_width / 2, 200, self.subtitle_text, font_mono, {1, 1, 1, 1}, 'centre'),
    }

    self.overlay = SettingsOverlay:new(resources, self, 300, 350, 680, 400)

    self.onMouseRelease = (function(self)
        self:setExit()
    end)

    self.onKeyRelease = (function(self)
        self:setExit()
    end)
end

-- Render this screen's contents.
function SettingsScreen:draw()
    love.graphics.clear(0, 0, 0, 1)

    -- UI parts
    for i in ipairs(self.ui) do
        self.ui[i]:draw()
    end

    -- Display the settings overlay.
    self.overlay:draw()

    if not self.fade:isDone() then
        love.graphics.setColor(unpack(self.fade:getColor()))
        love.graphics.rectangle('fill', 0, 0, gameState.scr_width, gameState.scr_height)
    end
end

-- Update the screen.
function SettingsScreen:update(dt)
    self.fade:update(dt)
end

return SettingsScreen

File: src/screens/SettingsScreen.py
# Skelly settings screen.
#
# By Chris Herborth (https://github.com/Taffer)
# MIT license, see LICENSE.md for details.

import pygame

from .ScreenBase import ScreenBase
from ..ui import SettingsOverlay
from ..ui import ColorFade

BLACK = pygame.Color('black')
BLACK_ALPHA = pygame.Color(BLACK.r, BLACK.g, BLACK.g, 0)  # BLACK, but fully transparent
WHITE = pygame.Color('white')


class SettingsScreen(ScreenBase):
    def __init__(self, game) -> None:
        super().__init__(game)

        self.next_screen = 'Journey'

        self.add_title()
        self.fade = ColorFade(BLACK, BLACK_ALPHA, 1)
        self.overlay = SettingsOverlay(game)

    def draw(self) -> None:
        self.game.surface.fill(BLACK)
        self.draw_title()

        self.game.manager.draw_ui(self.game.surface)

    def update(self, dt: float) -> None:
        self.fade.update(dt)

        if self.overlay.overlay_closed:
            self.can_exit = True

    def userevent(self, event: pygame.event.Event) -> None:
        self.overlay.userevent(event)
